src/symbols_linux.cpp

```diff
@@ -780,16 +767,65 @@ void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
 
         if (haveKernelSymbols()) {
             cc->sort();
-            array->add(cc);
+            mainArray->add(cc);
         } else {
             delete cc;
         }
     }
 
-    // In glibc, dl_iterate_phdr() holds dl_load_write_lock, therefore preventing
-    // concurrent loading and unloading of shared libraries.
-    // Without it, we may access memory of a library that is being unloaded.
-    dl_iterate_phdr(parseLibrariesCallback, array);
+    // Parse vm maps to find bounds.
+    CodeCacheArray array;
+    parseMaps(&array, mainArray->count());
+
+    int lib_count = array.count();
+    for (int i = 0; i < lib_count; i++) {
+        // We only parse libraries that were detected in parseMaps as loaded.
+        CodeCache* cc = array[i];
+        const char* image_base = (const char*)((const char*)cc->minAddress() - (const char*)cc->mapOffset());
+
+        // Do not parse same library more than once.
+        if (_elf_parsed_libraries.find(image_base) != _elf_parsed_libraries.end()) {
+            continue;
+        }
+
+        if (strchr(cc->name(), ':') != NULL) {
+            // Do not try to parse pseudofiles like anon_inode:name, /memfd:name
+        } else if (strcmp(cc->name(), "[vdso]") == 0) {
+            ElfParser::parseProgramHeaders(cc, (const char*)cc->minAddress(), (const char*)cc->maxAddress(), true);
+        } else {
+            // dlopen while parsing to prevent unloading the library and to make sure it is done loading.
+            // dlopen may fail opening the main executable and dynamic linker, which we ignore.
```

Comment:
But we can't ignore failures for other libraries, otherwise we may attempt to parse a library that has been just unloaded.
I'll fix this too as I'm refactoring this code anyway.